### PR TITLE
Enhancement: make a list to keep track of appointments

### PR DESCRIPTION
**Pull request by kyouscrap**

### CODE REVIEW

File: src/main/java/seedu/address/model/appointment/AppointmentContainsDatePredicate.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        if (person.getAppointment() == null) {
+            return false;
+        } else if (this.date == null) {
+            return true;
+        }
+
+        LocalDate startDate = person.getAppointment().getStart().toLocalDate();
+        LocalDate endDate = person.getAppointment().getEnd().toLocalDate();
+
+        return !this.date.isBefore(startDate) && !this.date.isAfter(endDate);
```
Comment (choonzies) [inline src/main/java/seedu/address/model/appointment/AppointmentContainsDatePredicate.java:33]: Just a nit, I think may be good to separate them out as different boolean statements first, for enhance readability and understandability. You may not have to take my suggestion though, as the isBefore and isAfter is relatively intuitive.
i.e
```java//
boolean predicateDateAfterStartDate = !this.date.isBefore(startDate);
boolean predicateDateBeforeEndDate = !this.date.isAfter(endDate);

// returns true when predicate date lies between appointment startDate and endDate
return predicateDateAfterStartDate && predicateDateBeforeEndDate;


```

File: src/test/java/seedu/address/model/appointment/AppointmentContainsDatePredicateTest.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+        // Non-matching dates
+        predicate = new AppointmentContainsDatePredicate(LocalDate.parse("01-01-2023", DATE_FORMATTER));
+        Person p = new PersonBuilder()
+                .withAppointment("Surgery",
+                        LocalDateTime.parse("23-10-2024-12-00", DATE_TIME_FORMATTER),
+                        LocalDateTime.parse("23-10-2024-14-00", DATE_TIME_FORMATTER))
+                .build();
+        assertFalse(predicate.test(p));
+    }
```
Comment (choonzies) [inline src/test/java/seedu/address/model/appointment/AppointmentContainsDatePredicateTest.java:113]: Can perhaps include another test case with the Predicate date being AFTER the appointment, to cover all partitions.

File: src/main/java/seedu/address/model/appointment/AppointmentContainsDatePredicate.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+public class AppointmentContainsDatePredicate implements Predicate<Person> {
+    private final LocalDate date;
+
+    public AppointmentContainsDatePredicate(LocalDate date) {
+        this.date = date;
+    }
+
+    @Override
+    public boolean test(Person person) {
+        if (person.getAppointment() == null) {
```
Comment (choonzies) [inline src/main/java/seedu/address/model/appointment/AppointmentContainsDatePredicate.java:24]: If person is null, will it throw a null exception?

File: src/test/java/seedu/address/model/appointment/AppointmentContainsDatePredicateTest.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+        // null -> returns false
+        assertFalse(firstPredicate.equals(null));
+
+        // different keywords but same field -> returns false
+        assertFalse(firstPredicate.equals(secondPredicate));
+    }
+
+    @Test
+    public void test_appointmentContainsDate_returnsTrue() {
```
Comment (choonzies) [inline src/test/java/seedu/address/model/appointment/AppointmentContainsDatePredicateTest.java:46]: I like your exhaustive test cases which follows the Boundary Value Analysis as well by testing the boundaries.

**Review by choonzies — COMMENTED** [review]
LGTM generally, can apply some defensive programming practices when implementing the predicate. Tiny nits regarding boolean statements and Law of Demeter.

The list and filtered list functions seems to work well, with good exhaustive test cases 😃

File: src/main/java/seedu/address/model/appointment/AppointmentContainsDatePredicate.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+            return false;
+        } else if (this.date == null) {
+            return true;
```
[the remark this answers is not in the document]
Comment (shuckycheese) [reply src/main/java/seedu/address/model/appointment/AppointmentContainsDatePredicate.java:27]: Noted! I will base the logic of schedule_all command on this implementation

File: src/main/java/seedu/address/model/appointment/AppointmentContainsDatePredicate.java (hunk 3 above)
Comment (shuckycheese) [reply src/main/java/seedu/address/model/appointment/AppointmentContainsDatePredicate.java:24]: I think if person is null, a null pointer exception will be thrown as you cannot call methods on null references. Consider adding an additional layer of checking to handle null cases?
